csrc/device_lower/utils.cpp

```diff
@@ -1013,6 +1016,775 @@ bool predicateAtEnd(ForLoop* loop) {
+std::string print(const Projection& proj);
+
+std::string print(const ValGroup& group) {
+  return group->toString();
+}
+
+std::string print(const PartOf<Projection>& part) {
+  auto str_or_null = [](Val* val) {
+    return val == nullptr ? "nullptr" : val->toInlineString();
+  };
+  return "PartOf(what=" + print(*part.what) +
+      ", inner_extent=" + str_or_null(part.inner_extent) +
+      ", selected_extent=" + str_or_null(part.selected_extent) + ")";
+}
+
+std::string print(const Composition<Projection>& vec) {
+  std::stringstream ss;
+  ss << "[";
+  for (const auto& g : vec) {
+    ss << print(g) << ", ";
+  }
+  ss << "]";
+  return ss.str();
+}
+
+std::string print(const std::monostate&) {
+  return "std::monostate";
+}
+
+std::string print(const Projection& proj) {
+  return Projection::dispatch(
+      [&](const auto& proj) { return print(proj); }, proj);
+}
+
+// Utilities to check if a ValGroup is contained in proj or its subtree.
+bool related(const Projection& proj, const ValGroup& to);
+
+bool related(const ValGroup& group, const ValGroup& to) {
+  return group == to;
+}
+
+bool related(const PartOf<Projection>& part, const ValGroup& to) {
+  return related(*part.what, to);
+}
+
+bool related(const Composition<Projection>& comp, const ValGroup& to) {
+  return std::any_of(
+      comp.begin(), comp.end(), [&](const auto& g) { return related(g, to); });
+}
+
+bool related(const std::monostate&, const ValGroup& to) {
+  return false;
+}
+
+bool related(const Projection& proj, const ValGroup& to) {
+  return Projection::dispatch(
+      [&](const auto& proj) { return related(proj, to); }, proj);
+}
+
+// Utilities to get the extent of a projection.
+Val* extent(const Projection& proj);
+
+Val* extent(const ValGroup& group) {
+  return group->front()->as<IterDomain>()->extent();
+}
+
+Val* extent(const PartOf<Projection>& part) {
+  return part.selected_extent;
+}
+
+Val* extent(const Composition<Projection>& comp) {
+  return std::accumulate(
+      comp.begin(),
+      comp.end(),
+      static_cast<Val*>(nullptr),
+      [](Val* acc, const auto& g) {
+        return SimplifyingIrBuilder::mulExpr(acc, extent(g));
+      });
+}
+
+Val* extent(const std::monostate&) {
+  NVF_ERROR(false, "Cannot get extent of std::monostate");
+}
+
+Val* extent(const Projection& proj) {
+  return Projection::dispatch(
+      [&](const auto& proj) { return extent(proj); }, proj);
+}
+
+// Simplify the abstract syntax tree so that it is easier to be pattern
+// matched.
+Projection simplify(Projection proj);
+
+Projection simplify(const ValGroup& group) {
+  return group;
+}
+
+PartOf<Projection> cancelCommonFactors(const PartOf<Projection>& part) {
+  // If `what` is a composition and inner_extent is a multiple of the extent of
+  // the last items in `what`, we can simplify the inner_extent and `what` by
+  // canceling the last items in `what`.
+  //
+  // Example:
+  // PartOf{what=[5,3,2], inner_extent=42} => PartOf{what=[5], inner_extent=7}
```

Comment:
Could you also mention how `PartOf{what=[5,3,2], inner_extent=42}` can be generated?

Reply:
![Screenshot_20240918_093052_Samsung Notes](https://github.com/user-attachments/assets/3512b496-5313-40ce-8d48-c968a6f93996)

Reply:
This is not a very good example. A better example could be
```
PartOf{what=[7,3,2], inner_extent=30}
```

Reply:
Do you mean a domain of extent 30 is split twice? I think I understand each split would generate `[1, 42], inner_extent=42` and `[5, 3, 2]`, but I don't follow why the could be combined.

Reply:
Yes, they split twice. They could combine because, for example, the `[1, 42]` is the loop domain, and the `[5, 3, 2]` is allocation domain.

Reply:
Hmm, I'm not sure why they could be combined. I thought this analysis is meant to understand what portion (subregion) an ID corresponds to in a domain. If that's the case, I don't understand what the combined domain of a loop domain and an allocation domain would mean.

Reply:
Oh, I think it is because the above transformation is mathematically equivalent to:
![Screenshot_20240918_122642_Samsung Notes](https://github.com/user-attachments/assets/ea699f42-197c-4058-bb72-d0c5d9e18293)

Reply:
Actually, I think I understand what you mean. 

1. Initial domain: `[1]`
2. Backward split by 42 -> `[30], inner_extent=42, selected_extent=1` 
3. Forward split by 6 -> `[5, 6], inner_extent=42, selected_extent=1`
4. Forward split by 2 -> `[5, 3, 2], inner_extent=42, selected_extent=1`

Am I right?

Reply:
Yes, right. I think I finally understand your question after you answered it yourself 😂.

Reply:
So, about this simplification, I suppose this is the same simplification we discussed earlier (`[5, 3]` -> `[5]`). In the case of `[5, 3, 2], inner_extent=42` -> `[5], inner_extent=7`, this means that we view the inner two IDs of `[3, 2]` as a non-modifiable unit, so it should be effectively equivalent to `[5], inner_extent=7`. However, what would happen if there's also an expr to traverse that uses `[3, 2]`? For example, let's say there's a split of `2` to `[1, 3]`. Since `what` no longer has `[2]`, this propagation would be just a no-op? Or, what about, e.g., a merge of `3` as the outer and `5` as the inner?

Reply:
This is a really good question. I love this question because it really touches the mathematical foundation of the proving we implemented.

When we start from `linear_g` and propagate to `5, 3, 2`, mathematically, we are proving that `linear_g ~ f(5, 3, 2)`, where `f` is a specific way to operates on its parameters, and `~` denote "equivalent" (e.g. `6 ~ [3, 2]` if you have `6 = merge(3, 2)`).

When we run the simplification, we simplify `f(5, 3, 2)` into `g(5)`, where `g` is another specific way to operate on its parameters. The process of simplifying `f(5, 3, 2)` into `g(5)` is just like in a math class, we simplify $f(x, y, z) = z + x^2 + 0 \cdot y - z$ as $g(x) = x^2$, where `y` and `z` is canceled means the value of `f(x, y, z)` does not depend on the value of `y` and `z`. Similarly, for this case, simplifying `f(5, 3, 2)` into `g(5)` means that, the index of `f(5, 3, 2)` does not depend on the index of `3` and `2`.

When we further propagate from `g(5)` to `domain`, mathematically, we are proving that `g(5) ~ h(x1, x2, ...)` where `x1`, `x2`, ... are ValGroups in `domain`. Note that this process only depend on the exprs between `5` and `domain`, nothing else. The split of `2` into `[1, 3]` is not on this path, so it will be unrelated. Merging `3` with `5` is related, but the only information that is related here is "3 is not a parameter of `g`", it does not matter where this `3` comes from and whether it is connected to `linear_g` or not, because here, we are just focusing on the subproblem of proving `g(5) ~ h(x1, x2, ...)` and `linear_g` is unrelated to this subproblem.

Reply:
So, IIUC, what you're saying is that no matter what other exprs are traversed, if it uses anything other than `5`, it should be just ignored as it shouldn't matter. Is my understanding correct?

Reply:
I think what I missed was this part:

> inner_extent is a multiple of the extent of the last items in `what`,

So, because of this condition, the canceled factor should not matter for the remaining traversal. 

It finally makes sense to me. Thanks for the updated comment. That helped a lot.